misk-feature/src/main/kotlin/misk/feature/FeatureFlags.kt

```diff
@@ -0,0 +1,74 @@
+package misk.feature
+
+/**
+ * Interface for evaluating feature flags.
+ */
+interface FeatureFlags {
+
+  /**
+   * Calculates the value of an boolean feature flag for the given token and attributes.
+   * @see [getEnum] for param details
+   */
+  fun getBool(
+    feature: Feature,
+    token: String,
+    attrs: Attrs? = null
+  ): Boolean
+
+  /**
+   * Calculates the value of an integer feature flag for the given token and attributes.
+   * @see [getEnum] for param details
+   */
+  fun getInt(
+    feature: Feature,
+    token: String,
+    attrs: Attrs? = null
+  ): Int
+
+  /**
+   * Calculates the value of a string feature flag for the given token and attributes.
+   * @see [getEnum] for param details
+   */
+  fun getString(
+    feature: Feature,
+    token: String,
+    attrs: Attrs? = null
+  ): String
+
+  /**
+   * Calculates the value of an enumerated feature flag for the given token and attributes.
+   * @param feature name of the feature flag to evaluate.
+   * @param token unique primary token for the entity the flag should be evaluated against.
+   * @param default default value to return if there was an error evaluating the flag or the flag
+   *   does not exist.
+   * @param attrs additional attributes to provide to flag evaluation.
+   */
+  fun <T : Enum<T>> getEnum(
+    feature: Feature,
+    token: String,
+    clazz: Class<T>,
+    attrs: Attrs? = null
+  ): T
+}
+
+inline fun <reified T : Enum<T>> FeatureFlags.getEnum(
+  feature: Feature,
+  token: String,
+  attrs: Attrs? = null
+): T = getEnum(feature, token, T::class.java, attrs)
+
+/**
+ * Typed feature string.
+ */
+data class Feature(val name: String)
+
+/**
+ * Extra attributes to be used for evaluating features.
+ */
+data class Attrs(
+  val stringAttrs: Map<String, String> = mapOf(),
```

Comment:
Do we want to pass this through exactly as-is? Do we have a use case? This seems like unnecessary complexity

Reply:
This is how you'd pass user properties through for segmenting them, like by country, platform (ios/android)

Reply:
Gotcha. Can you add that example? Should we use types for this?

Reply:
I'm not attached to this. I think almost all usages will be passing in a string map.

Not sure what you mean by using types. Can you expand on that?

Reply:
In my experience if you give developers a `Map<String, String>` then some of them will use `state/prov: CALIFORNIA` and others will use `STATE: ca`. Giving them a type instead makes it easier to be consistent, which increases the value of the data gathered.

Here’s a hypothetical sketch:

```
/** Subtypes are data classes whose fields are captured with feature flag lookups. */
interface Attributes {
}

data class PlasmaAttributes {
  val state: UspsState,
  val statusLevel: StatusLevel,
  val userAgent: UserAgent
} : Attributes
```

used like this:

```
  override fun getBoolean(feature: Feature, token: String, attributes: Attributes) Boolean {
    ...
  }
```

That way every developer says “what is an Attributes?” and they find their project’s implementation and that makes things typesafe.

Lets land this as-is before adding that though.

Reply:
We want to have consistency across services. I'd like to explore deeper how to do that in a type safe manner. My initial thought is that we may get most of what we want by exposing constants for common keys rather than adding types